src/python/foglamp_daemon.py

```diff
@@ -26,31 +48,121 @@ def do_something(logf):
     logger.addHandler(file_handler)
     logger.setLevel(logging.DEBUG)
 
-    start()
+    # The main daemon process
+    start_controller()
 
 
-def start_daemon(pidf, logf, wdir):
+def run():
+    """
+    Run foglamp server in foreground
+    :return:
     """
-    Launches the daemon
 
-    :param pidf: pidfile
-    :param logf: log file
-    :param wdir: working directory
+    if is_running():
+        message = "Daemon already running. Check PID %s.\n"
+        sys.stderr.write(message % get_pid())
+        return is_running()
+
+    logging.basicConfig(level=logging.DEBUG)
+    logging.getLogger("foglamp").setLevel(logging.DEBUG)
+
+    start_controller()
+
+
+def start():
+    """
+    Launches the daemon
     """
 
-    # XXX: pidfile is a context
+    # Get the pid from the pidfile
+    pid = get_pid()
+
+    if pid is not None:
+        message = "Daemon already running under PID %s.\n"
+        sys.stderr.write(message % pid)
+        return is_running()
+
     with daemon.DaemonContext(
         working_directory=wdir,
         umask=0o002,
         pidfile=pidfile.TimeoutPIDLockFile(pidf)
     ) as context:
         do_something(logf)
 
+def stop():
+    """
+    Stop the daemon
+    """
+
+    # Get the pid from the pidfile
+    pid = get_pid()
+
+    if pid is None:
+        message = "pidfile %s does not exist. Daemon not running?\n"
+        sys.stderr.write(message % pidf)
+        return is_running()
+
+    # Try killing the daemon process
+    try:
+        while True:
+            os.kill(pid, signal.SIGTERM)
+            time.sleep(0.1)
+    except OSError as err:
+        err = str(err)
+        if err.find("No such process") > 0:
+            if os.path.exists(os.path.expanduser(pidf)):
+                os.remove(os.path.expanduser(pidf))
+        else:
+            sys.stdout.write(str(err))
+            sys.exit(1)
+
+    return is_running()
+
+
+def restart():
+    """
+    Relaunches the daemon
+
+    :param pidf: pidfile
+    :param logf: log file
+    :param wdir: working directory
+    """
+
+    if is_running():
+        stop()
+    start()
+
+    return is_running()
+
+
+def is_running():
+    """
+    Check if the daemon is running.
+    """
+
+    return get_pid() is not None
+
+
+def get_pid():
+    """
+    Returns the PID from pidf
+    """
+
+    try:
+        pf = open(pidf,'r')
```

Comment:
Cannot. with statement does not catch exception. It is not allowed https://mail.python.org/pipermail//python-ideas/2011-November/012925.htmlhttps://mail.python.org/pipermail//python-ideas/2011-November/012925.html.